Pagination, sorting, filtering

cmurphy opened the pull request:
Commit summary:

### Add unit tests for partition listing

### Refactor proxy store

Filtering and sorting needs to operate on unstructured data. It also
needs to operate after the parallel partitioner, higher in the store
stack. This means that the proxy Store needs to return raw, unstructured
data up to the partitioner. This change moves all conversions from
unstructured Kubernetes types to apiserver objects up from the proxy
store into the partitioner.

### Add filtering to partition store

Extend the partition store to parse query params as list filters. Filter
keys use dot notation to denote the subfield of an object to filter on.
Example:

GET /v1/secrets?filter=metadata.name=foo

Filters are ANDed together, so an object must match all filters to be
included in the list. Example:

GET /v1/secrets?filter=metadata.name=foo&filter=metadata.namespace=bar

Arrays are searched for matching items. If any item in the array
matches, the item is included in the list. Example:

GET /v1/pods?filter=spec.containers.image=alpine

### Add sorting to partition store

Extend the partition store to parse the "sort" query parameter as a
sorting condition. Dot notation is used to denote the object field.
Preceding the key with "-" denotes descending (reverse) order.

Example sorting by name:

GET /v1/secrets?sort=metadata.name

Reverse sorting by name:

GET /v1/secrets?sort=-metadata.name

All values are converted to strings and sorted lexicographically.

### Add pagination to partition store

Extend the partition store to parse page and pagesize parameters and
return a subset of list results. The total number of pages is included
in the list response.

Request an initial page:

GET /v1/secrets?pagesize=10

or

GET /v1/secrets?pagesize=10&page=1

Request subsequent pages, or arbitrary pages:

GET /v1/secrets?pagesize=10&page=37

If a page number is out of bounds, an empty list is returned.

### Add caching to pagination

Cache filtered, sorted results for fast subsequent page retrieval.

Requests for cached queries need to include the list revision number
along with other queries. If no specific revision is requested, a new
fetch is done in order to get the latest revision. The revision is
included in the list response.

Example first request:

GET /v1/secrets?pagesize=10

Example subsequent page request:

GET /v1/secrets?pagesize=10&page=1&revision=107740

### Add pagination tests for partition listing

### Add secondary sort parameter

Extend the sorting functionality in the partition store to support
primary and secondary sorting criteria. Sorting parameters are specified
by a single 'sort' query and comma separated. Example:

Sort by name and creation time:

GET /v1/secrets?sort=metadata.name,metadata.creationTimestamp

Reverse sort by name, normal sort by creation time:

GET /v1/secrets?sort=-metadata.name,metadata.creationTimestamp

Normal sort by name, reverse sort by creation time:

GET /v1/secrets?sort=metadata.name,-metadata.creationTimestamp

### Add README
Add an initial README to explain the basics of the API and the new
parameters.

Issue: https://github.com/rancher/rancher/issues/38427

~Depends on https://github.com/rancher/steve/pull/60~
Depends on https://github.com/rancher/steve/pull/63

cmurphy commented on go.mod:
```diff
@@ -6,6 +6,7 @@ replace (
 	github.com/crewjam/saml => github.com/rancher/saml v0.0.0-20180713225824-ce1532152fde
 	github.com/knative/pkg => github.com/rancher/pkg v0.0.0-20190514055449-b30ab9de040e
 	github.com/matryer/moq => github.com/rancher/moq v0.0.0-20190404221404-ee5226d43009
+	github.com/rancher/apiserver => github.com/cmurphy/apiserver v0.0.0-20221110205644-da6f6ae1675a
```
DO NOT MERGE without updating this

MbolotSuse commented on README.md:
```diff
@@ -0,0 +1,144 @@
+steve
```
Thanks for adding docs.

cmurphy commented:
Latest changes - 
1. updated the pagination cache key to use accesscontrol.AccessFor instead of username and groups
2. changed the default cache size to 1000 and made it configurable

rmweir commented on pkg/stores/partition/store.go:
```diff
@@ -2,28 +2,85 @@ package partition
+	return &Store{
+		Partitioner: partitioner,
+		asl:         asl,
+		listCache:   cache.NewLRUExpireCache(cacheSize),
```
I think ~~the max size should be configurable and~~ that it should be possible to disable the use of the cache entirely. It could be hard to come up with the right number because the size of lists will differ so much depending on object type and user permissions. What if a user's permissions have changed since the last cache? I think the 30 minute expiry should be fine for this but maybe someone from security can approve.

cmurphy replied:
> it should be possible to disable the use of the cache entirely

done, added a test for this

> What if a user's permissions have changed since the last cache?

This is done now with the access set check, the access set ID is part of the cache key and is recalculated on every request.

KevinJoiner replied:
I see you added an ENV check for disabling the cache, but would it also be beneficial to use a query parameter to bypass the cache per request? This would allow the UI to use the cache based on the resource type.

cmurphy commented on pkg/stores/partition/listprocessor/processor.go:
```diff
@@ -0,0 +1,287 @@
+	if sortKeys != "" {
+		sortParts := strings.SplitN(sortKeys, ",", 2)
+		primaryField := sortParts[0]
+		if primaryField[0] == '-' {
```
Not quite. The length of the parameter is checked on 120. Then the slice created on 121 must have at least one non-empty string in it. That string is shortened by 1 on 125. At that point, it's possible it could be empty, hence the check on 127.

KevinJoiner replied:
You will still need the check for both primaryField and secondaryField because of the `strings.SplitN(...)`
https://go.dev/play/p/ELlnx4eMRwS

cmurphy commented on README.md:
```diff
@@ -0,0 +1,144 @@
+* `/v1/{type}` - all cluster-scoped resources OR all resources in all
+  namespaces of type `{type}` that the user has access to
+* `/v1/{type}/{name}` - cluster-scoped resource of type `{type}` and unique name `{name}`
+* `/v1/{type}/{namespace} - all resources of type `{type}` under namespace `{namespace}`
```
:+1:

cmurphy commented on README.md:
```diff
@@ -0,0 +1,144 @@
+steve supports query parameters to perform actions or process data on top of
+what Kubernetes supports.
+
+#### `link`
+
+Trigger a link handler, which is registered with the schema. Examples are
+calling the shell for a cluster, or retrieving logs during catalog operations.
+
+#### `action`
+
+Trigger an action handler, which is registered with the schema. Examples are
+generating a kubeconfig for a cluster, or installing an app from a catalog.
```
:+1:

cmurphy commented on pkg/stores/partition/listprocessor/processor.go:
```diff
@@ -0,0 +1,287 @@
+		return false
+	}
+	switch typedVal := objValue.(type) {
+	case string, int, bool:
```
Fixed, added a test to cover this

KevinJoiner commented on README.md:
```diff
@@ -0,0 +1,153 @@
+steve
+=====
+
+steve is a lightweight API proxy for kubernetes whose aim is to create an
```
```suggestion
Steve is a lightweight API proxy for Kubernetes whose aim is to create an
```

KevinJoiner commented on README.md:
```diff
@@ -0,0 +1,153 @@
+
+steve is a lightweight API proxy for kubernetes whose aim is to create an
+interface layer suitable for dashboards to efficiently interact with
+kubernetes.
```
```suggestion
Kubernetes.
```

KevinJoiner commented on README.md:
```diff
@@ -0,0 +1,153 @@
+### /v1 API
+
+steve registers all Kubernetes resources as schemas in the /v1 API. Any
+endpoint can support methods GET, POST, PATCH, PUT, or DELETE depending on what
```
```suggestion
endpoint can support methods GET, POST, PATCH, PUT, or DELETE, depending on what
```

KevinJoiner commented on README.md:
```diff
@@ -0,0 +1,153 @@
+
+### Query parameters
+
+steve supports query parameters to perform actions or process data on top of
```
```suggestion
Steve supports query parameters to perform actions or process data on top of
```

KevinJoiner commented on README.md:
```diff
@@ -0,0 +1,153 @@
+#### `link`
+
+Trigger a link handler, which is registered with the schema. Examples are
+calling the shell for a cluster, or following logs during cluster or catalog
```
```suggestion
calling the shell for a cluster, or following logs during a cluster or catalog
```

KevinJoiner replied:
You are right, I misread "operations".

KevinJoiner commented on README.md:
````diff
@@ -0,0 +1,153 @@
+POST /v1/catalog.cattle.io.clusterrepos/rancher-partner-charts?action=install
+```
+
+#### `filter` (new for Rancher v2.7.x)
````
We could leave this off since Steve is used outside of Rancher.
```suggestion
#### `filter` 
```

KevinJoiner commented on README.md:
```diff
@@ -0,0 +1,153 @@
+
+Only applicable to list requests (`/v1/{type}` and `/v1/{type}/{namespace}`)
+
+Results can be sorted lexicographically by primary and secondary colums.
```
```suggestion
Results can be sorted lexicographically by primary and secondary columns.
```

KevinJoiner commented on README.md:
````diff
@@ -0,0 +1,153 @@
+/v1/{type}?sort=-metadata.name
+```
+
+Secondary sort criteria is comma separated.
````
```suggestion
The secondary sort criteria is comma separated.
```

KevinJoiner commented on README.md:
```diff
@@ -0,0 +1,153 @@
+
+Results can be batched by pages for easier display.
+
+Example initial request, returns a page with 10 results:
```
```suggestion
Example initial request returns a page with 10 results:
```

KevinJoiner replied:
We can use a semicolon without a conjunction if they are independent clauses.
```suggestion
Example initial request; returns a page with 10 results.
```

cmurphy replied:
Changed to "returning" to make it one clause and remove the comma

KevinJoiner commented on pkg/stores/partition/store_test.go:
```diff
@@ -0,0 +1,1832 @@
+package partition
```
Nothing to Change here; just a suggestion for future tests. If you leverage `testify.Suites` It is easier to break up tests into separate functions.

KevinJoiner reviewed (changes requested):
Nice job on the implementation. The separate commits made it easy to follow.

KevinJoiner reviewed (changes requested):
Also, can you add package comments?

cmurphy commented on pkg/stores/partition/store.go:
```diff
@@ -175,17 +300,79 @@ func (s *Store) Watch(apiOp *types.APIRequest, schema *types.APISchema, wr types
-	limitString := req.URL.Query().Get("limit")
-	limit, err := strconv.Atoi(limitString)
+func toAPI(schema *types.APISchema, obj runtime.Object) types.APIObject {
+	if obj == nil || reflect.ValueOf(obj).IsNil() {
```
> I think between these two cases, it is better to assume that the runtime.Object's underlying value is a non-nil struct, than it is to assume that the runtime.Object's underlying value is a struct with a pointer receiver mainly because the latter is not idiomatic behavior in go.

runtime.Object should basically always contain a struct with methods defined for pointer receivers because the types that implement [runtime.Object](https://pkg.go.dev/k8s.io/apimachinery/pkg/runtime#Object) include [metav1.TypeMeta](https://pkg.go.dev/k8s.io/apimachinery/pkg/apis/meta/v1#TypeMeta.GetObjectKind) as an [embedded field](https://github.com/rancher/rancher/blob/38979f8ebe5c1ecb0121494d185758741fcb0d5b/pkg/apis/management.cattle.io/v3/public_authn_types.go#L12) and have autogenerated [DeepCopyObject pointer methods](https://github.com/rancher/rancher/blob/38979f8ebe5c1ecb0121494d185758741fcb0d5b/pkg/apis/management.cattle.io/v3/zz_generated_deepcopy.go#L539). So I think we still need reflect no matter what. We could add safety by checking `.Kind` before checking `.IsNil()`.

cmurphy commented:
> Also, can you add package comments?

![giphy](https://user-images.githubusercontent.com/1220710/205409556-e0c580cc-81f4-4996-abbf-f202ebd8873b.gif)

KevinJoiner commented on pkg/stores/partition/store.go:
```diff
@@ -88,38 +158,96 @@ func (s *Store) List(apiOp *types.APIRequest, schema *types.APISchema) (types.AP
-	list, err := lister.List(apiOp.Context(), limit, resume)
+	key, err := s.getCacheKey(apiOp, opts)
 	if err != nil {
 		return result, err
 	}
 
-	for items := range list {
-		result.Objects = append(result.Objects, items...)
+	var list []unstructured.Unstructured
```
I can agree with that sentiment. I can be swayed either way so let's just leave it as is for now.

KevinJoiner commented on pkg/stores/partition/listprocessor/processor.go:
```diff
@@ -0,0 +1,300 @@
+// getLimit extracts the limit parameter from the request or sets a default of 100000.
+// The default limit can be explicitly disabled by setting it to a negative value.
+// If the default is accepted, clients must be aware that the list may be incomplete, and use the "continue" token to get the next chunk of results.
+func getLimit(apiOp *types.APIRequest) int {
```
We should either update our behavior to state that if the limit is 0 or negative we unset the limit parameter or leave the check that overrides `0`to be the `defaultLimit`. I couldn't find any official documentation on `limit=0` in the kubeapi references, but through some simple curl testing, if I add  `limit=0` I get back all the results. This could be unexpected for users since our previous implementation would override the limit if it was set to '0'

KevinJoiner commented on pkg/stores/partition/parallel.go:
```diff
@@ -87,6 +87,9 @@ func (p *ParallelPartitionLister) List(ctx context.Context, limit int, resume st
 			limit = state.Limit
 		}
 	}
+	if revision != "" {
+		state.Revision = revision
+	}
```
Do we need this check since `state.Revision` is already an empty string?

KevinJoiner commented on pkg/stores/partition/store.go:
```diff
@@ -184,11 +187,14 @@ func (s *Store) List(apiOp *types.APIRequest, schema *types.APISchema) (types.AP
 			return result, err
 		}
 		list = listprocessor.FilterList(stream, opts.Filters)
+		if lister.Err() != nil {
```
Maybe a comment here stating that we are checking for any errors returned during the listing operation. And that FilterList guarantees the stream has been consumed. 
I just have the fear of things getting moved around since there is no clear dependency on `FilterList(...)`

cmurphy replied:
yeah good point, it wasn't even obvious to me at first why I couldn't check the error first

rmweir reviewed (approved):
LGTM

cmurphy commented:
Updated to include merged apiserver change
